### PR TITLE
Upgrade GitHub Actions

### PR DESCRIPTION
**Pull request by cclauss**
Also, remove redundant settings.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -14,51 +14,27 @@ jobs:
...
         allow-prereleases: true
-    - name: Set up development Python ${{ matrix.python-version }}
-      if: endsWith(matrix.python-version, '-dev')
-      uses: deadsnakes/action@v2.1.1
-      with:
-        python-version: ${{ matrix.python-version }}
     - name: Python (Install Dependencies)
       run: |
-        python -m pip install --upgrade pip setuptools tox coverage build
+        python -m pip install --upgrade pip
```
Comment (facelessuser) [inline .github/workflows/build.yml:36]: This seems like an unnecessary change. If any of these are installed, ensuring we have the latest is not a problem, if they are not installed, they will still be installed.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -97,7 +73,7 @@ jobs:
       env:
         TOX_SPELL_REQUIRE: both
       run: |
-        python -m tox
```
Comment (facelessuser) [inline .github/workflows/build.yml:100]: I do not wish to sequence it in this way. I'd like each test to run and fail independently.

**Review by facelessuser — COMMENTED** [review]
While there may be some good things in here, there are a lot of large, opinionated decisions being made about how we should perform our actions in this PR. I would have appreciated a discussion first about the proposed changes as that would have saved you some time.

I don't mind upgrading actions, I also don't mind removing deadsnakes. I'm also not against bumping `max-parallel`, but I don't think I'm interested in the other changes.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -14,51 +14,27 @@ jobs:
...
-    env:
-      TOXENV: ${{ matrix.tox-env }}
-
     runs-on: ${{ matrix.platform }}
 
     steps:
-    - uses: actions/checkout@v3
+    - uses: actions/checkout@v4
     - name: Set up Python ${{ matrix.python-version }}
-      if: "!endsWith(matrix.python-version, '-dev')"
```
Comment (cclauss) [inline .github/workflows/build.yml:49]: None of the `matrix.python-version` values end in `-dev` so this test is not needed.

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -14,51 +14,27 @@ jobs:
...
       uses: actions/setup-python@v4
       with:
         python-version: ${{ matrix.python-version }}
         allow-prereleases: true
-    - name: Set up development Python ${{ matrix.python-version }}
-      if: endsWith(matrix.python-version, '-dev')
```
Comment (cclauss) [inline .github/workflows/build.yml:55]: None of the `matrix.python-version` values end in `-dev` so deadsnakes will never be used.

File: .github/workflows/build.yml (hunk 2 above)
Comment (cclauss) [reply .github/workflows/build.yml:100]: Each will run separately and fail independently.  That is why we use a matrix.

File: .github/workflows/build.yml (hunk 3 above)
Comment (facelessuser) [reply .github/workflows/build.yml:49]: Again, this is part of deadsnakes (when we use it), but as I stated, I am okay removing this as now we can utilize prereleases.

File: .github/workflows/build.yml (hunk 4 above)
Comment (facelessuser) [reply .github/workflows/build.yml:55]: Part of deadsnakes, we can remove this.

**Comment by cclauss** [comment]
Why is Python 3.11 not a required test?

Screenshot 2023-09-10 at 15 56 53

File: .github/workflows/build.yml (hunk 1 above)
Comment (cclauss) [reply .github/workflows/build.yml:36]: There is no need to upgrade pip if you have already installed your dependencies on the old version.  See the GitHub Actions examples.

**Comment by facelessuser** [comment]
Originally, I was trying out the "required" tests option. I always forget to update new tests and remove old tests. I don't have non-required tests, so I will remove the "required" settings.

File: .github/workflows/build.yml (hunk 2 above)
Comment (facelessuser) [reply .github/workflows/build.yml:100]: I originally had it done this way so that I could independently control `continue-on-error`. Sometimes I may attempt to add support for a future version, but dependencies or Python versions can be buggy in early stages (beta/alpha).

**Comment by cclauss** [comment]
Added `if: matrix.tox-env == 'documents'` then install Aspell

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -100,6 +100,10 @@ jobs:
       uses: actions/setup-python@v4
       with:
         python-version: 3.x
+    - name: Install Aspell
```
Comment (facelessuser) [inline .github/workflows/build.yml:103]: Linting doesn't need to install Aspell, that is why I prefer to have it as a separate action. I think it is more clear when it is separate.

**Comment by facelessuser** [comment]
I guess I am okay with docs and lint as they do appear separate now.

I do need `continue-on-error` easily configurable. The current setup does not allow this and I consider it a regression.

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -14,51 +14,30 @@ jobs:
...
-            tox-env: py310
-            continue-on-error: False
-          - python-version: '3.11'
-            tox-env: py311
-            continue-on-error: False
-          - python-version: '3.12'
-            tox-env: py312
-            continue-on-error: False
+          - python-version: 3.12
+            continue-on-error: false
```
Comment (cclauss) [inline .github/workflows/build.yml:23]: This can be removed when we are confident about Python 3.12 release candidates or when it is production released in three weeks.

**Comment by facelessuser** [comment]
Yeah, as soon as you include one, the tests don't run unless they are all defined, which is why I just define them all.

File: .github/workflows/build.yml (hunk 6 above)
Comment (facelessuser) [reply .github/workflows/build.yml:23]: Sorry, see my global comment below. Tests don't run if only one is defined. I don't know why, but it is just how GitHub works.

File: .github/workflows/build.yml (hunk 1 above)
Comment (facelessuser) [reply .github/workflows/build.yml:36]: My point generally was that I wanted to ensure all the listed packages were included as the latest, that includes `setuptools`, `tox`, `coverage`, `build`. Admittedly, some of these are never on the system, so they just install as normal, but some, like `setuptools` often is, and I want to ensure the latest is there.

That is why I don't understand why we want to only upgrade pip, just upgrade them all, and if they are already upgraded, they wont update.

File: .github/workflows/build.yml (hunk 6 above)
Comment (facelessuser) [reply .github/workflows/build.yml]: Please add back all of these. I'm just going to have to do it later when I need to alter the setup.

```yaml
        include:
          - python-version: 3.8
            tox-env: py38
            continue-on-error: False
          - python-version: 3.9
            tox-env: py39
            continue-on-error: False
          - python-version: '3.10'
            tox-env: py310
            continue-on-error: False
          - python-version: '3.11'
            tox-env: py311
            continue-on-error: False
          - python-version: '3.12'
            tox-env: py312
            continue-on-error: False
```

**Comment by facelessuser** [comment]
Thanks for your patience, I think everything looks okay now.

@gir-bot lgtm
